### PR TITLE
Try to support standard `http_proxy` environment variable

### LINKED ISSUE #1682: [Enhancement] Support http_proxy
* `carthage version`:  0.18.1
* `xcodebuild -version`:  8C1002
* Are you using `--no-build`? No
* Are you using `--no-use-binaries`? No
* Are you using `--use-submodules`? No

Carthage can't use proxy server to download prebuilt binaries.
It's a much needed feature for developers in China.

### PR DESCRIPTION
**Pull request by ikesyo**
Would resolve #1682.

### CODE REVIEW

**Comment by mdiep** [comment]
Any suggestion for how to test this? I might be able to check this at work this week, but I'm not sure how to check from my personal machine.

**Comment by tspike** [comment]
any chance we can get this merged? currently a blocker for us

**Comment by bartoszj** [comment]
Hi. Any progress with this task? A proxy setting could be tested with [Charles Proxy](https://www.charlesproxy.com/). Also, it would be nice to add the proxy setting to binary framework downloads.

**Comment by pnull** [comment]
👍

File: Source/CarthageKit/GitHub.swift [hunk 1]
```diff
@@ -145,14 +145,46 @@ extension Client {
...
+
+			if let port = components.port {
+				dictionary[kCFNetworkProxiesHTTPPort as AnyHashable] = port
+				dictionary[kCFNetworkProxiesHTTPSPort as AnyHashable] = port
+				components.port = nil
+			}
+			dictionary[kCFNetworkProxiesHTTPEnable as AnyHashable] = true
+			dictionary[kCFNetworkProxiesHTTPSEnable as AnyHashable] = true
+			dictionary[kCFNetworkProxiesHTTPProxy as AnyHashable] = components.url
+			dictionary[kCFNetworkProxiesHTTPSProxy as AnyHashable] = components.url
```
Comment (mdiep) [inline Source/CarthageKit/GitHub.swift]: I don't think we should be configuring HTTP and HTTPS with the same proxy. We should set them from the respective environment variables.

**Comment by mdiep** [comment]
Finally wired up Charles to test this out. Sorry it took me so long!

I think we should treat HTTP/HTTPS separately, but other than that this looks good!

**Comment by xiongmao86** [comment]
@ikesyo Could you support socks5 proxy as well?

**Comment by xiongmao86** [comment]
Also has a question, The team replied that they don’t support http way of downloading git repo, why setting http proxy?

File: Source/CarthageKit/GitHub.swift (hunk 1 above)
Comment (jakeadams) [reply Source/CarthageKit/GitHub.swift]: I think a `Proxy` type would be good to have so it can also be testable.

**Comment by blender** [comment]
This has been improved and superseded by #2371 . Closing. 

Thanks everyone for the effort and the valuable contributions.

**Pull request closed without merging**